src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,175 @@
+            {
+                var project = document.Project;
+                var typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+
+                // Find all namespaces in scope at current cursor location, 
+                // which will be used to filter so the provider only returns out-of-scope types.
+                var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
+                Action<CompletionItem> handleAccessibleItem = item => AddItems(item, completionContext, namespacesInScope);
+
+                // Get completion items from current project.
+                await typeImportCompletionService.GetAccessibleTopLevelTypesFromProjectAsync(project, handleAccessibleItem, cancellationToken)
+                    .ConfigureAwait(false);
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var reference in compilation.References)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var declarationsInReference = ImmutableArray<CompletionItem>.Empty;
+                    if (reference is CompilationReference compilationReference)
+                    {
+                        // Get completion items from source references.
+                        await typeImportCompletionService.GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                            project.Solution,
+                            compilation,
+                            compilationReference,
+                            handleAccessibleItem,
+                            cancellationToken).ConfigureAwait(false);
+                    }
+                    else if (reference is PortableExecutableReference peReference)
+                    {
+                        // Get completion items from metadata references.
+                        typeImportCompletionService.GetAccessibleTopLevelTypesFromPEReference(
+                            project.Solution,
+                            compilation,
+                            peReference,
+                            handleAccessibleItem,
+                            cancellationToken);
+                    }
+                }
+            }
+
+            return;
+
+            static void AddItems(CompletionItem item, CompletionContext completionContext, ImmutableHashSet<string> namespacesInScope)
+            {
+                var containingNamespace = TypeImportCompletionItem.GetContainingNamespace(item);
+                if (!namespacesInScope.Contains(containingNamespace))
+                {
+                    completionContext.AddItem(item);
+                }
+            }
+        }
+
+        private ImmutableHashSet<string> GetNamespacesInScope(Document document, SyntaxContext syntaxContext, CancellationToken cancellationToken)
```

Comment:
Immutable hash collection are actually pretty high overhead and high CPU hit.  Literally just use a pooledhashset for best result.

Reply:
Interesting. But I can't set the comparer on PooledHashSet, so I can't use that. Would regular HashSet any better?

Reply:
yes, much.  and you can make your own pool :)  in your case you'd need two pools one of teh casesensitive sets and one for the insensitive ones. it's not hard to just use ObjectPool and its siblings directly.

Reply:
OK, I will just use regular hashset for now. And switch to use PooledObject if it end up showing in the perf trace.

Reply:
(but how can immutable version be less efficient than the mutable one?)

Reply:
happy to chat on gitter aobut that :)